runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -83,6 +83,79 @@ isPolymorphicMHMethod(J9JavaVM *vm, J9Class *declaringClass, J9UTF8 *methodName)
 	return false;
 }
 
+/**
+ * @brief Add a MemberName to the list of MemberNames for the J9Class that will
+ * correspond to its clazz field.
+ *
+ * This must be done immediately prior to initializing vmtarget.
+ *
+ * clazzObject must be the value of the MemberName's clazz field, or the value
+ * that will be assigned to clazz immediately upon success.
+ *
+ * On error, the current exception will be set:
+ * - to OutOfMemoryError for allocation failure.
+ *
+ * The caller must have VM access.
+ *
+ * @param[in] currentThread the J9VMThread of the current thread
+ * @param[in] memberNameObject the MemberName object to add to the list
+ * @param[in] clazzObject the value of memberNameObject.clazz
+ * @return true for success, or false on error
+ */
+static bool
+addMemberNameToClass(J9VMThread *currentThread, j9object_t memberNameObject, j9object_t clazzObject)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9Class *j9clazz = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, clazzObject);
+
+	jobject weakRef = vmFuncs->j9jni_createGlobalRef((JNIEnv*)currentThread, memberNameObject, JNI_TRUE);
+
+	omrthread_monitor_enter(vm->memberNameListsMutex);
```

Comment:
I was asked to use a pool to allocate the list nodes. Since the pool needs a global mutex anyway, I used it to protect the lists as well. Outside of STW, the critical section will usually be constant-time, only taking longer if pruning is necessary